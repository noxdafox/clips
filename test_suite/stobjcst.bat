(load "stobjcst.clp")
(list-defrules)
(describe-class BASE)
(describe-class DERIVE1)
(describe-class DERIVE2)
(describe-class DERIVE3)
(clear)
(defclass A (is-a USER)
  (slot foo (cardinality 3 4)))
(defclass A (is-a USER)
  (slot foo (type SYMBOL) (range 3 4)))
(defclass A (is-a USER)
  (slot foo (min-number-of-elements 3)))
(defclass A (is-a USER)
  (slot foo (allowed-words abc)))
(defclass A (is-a USER)
  (slot foo (allowed-values)))
(defclass A (is-a USER)
  (slot foo (allowed-integers abc)))
(defclass A (is-a USER)
   (slot foo (type INTEGER) (default abc)))
(defclass A (is-a USER)
   (slot foo (type INTEGER) (default-dynamic abc)))
(set-static-constraint-checking FALSE)
(defclass A (is-a USER)
   (slot foo (type INTEGER) (default abc)))
(defclass B (is-a USER)
   (slot foo (type INTEGER) (default-dynamic abc)))
(set-static-constraint-checking TRUE)
(clear)
