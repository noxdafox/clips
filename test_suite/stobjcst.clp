(defclass FOO (is-a USER) (role concrete) (pattern-match reactive) 
   (slot x (type SYMBOL) (create-accessor ?NONE))
   (slot y (type INTEGER)))

(defclass BAR (is-a USER) (role concrete) (pattern-match reactive) 
   (slot a (type SYMBOL INTEGER) (create-accessor ?NONE))
   (slot b (type INTEGER FLOAT) (create-accessor ?NONE))
   (slot c (type SYMBOL STRING) (create-accessor ?NONE))) 

(defrule error-1
  (object (is-a FOO) (x ?x) (y ?x))
  =>)

(defrule error-2
  (object (is-a FOO) (x ?x))
  (object (is-a FOO) (y ?x))
  =>)

(defrule error-3
  (object (is-a BAR) (a ?x) (b ?x) (c ?x))
  =>)

(defrule error-4
  (object (is-a BAR) (a ?x))
  (object (is-a BAR) (b ?x))
  (object (is-a BAR) (c ?x))
  =>)
    
(defrule okay!-1 
  (object (is-a FOO) (x ?x) (y ?y))
  (object (is-a BAR) (a ~?x&~?y))
  =>)

(defclass YAK (is-a USER) (role concrete) (pattern-match reactive)  
   (slot a (type INTEGER) (range 4 7) (create-accessor ?NONE)))

(defclass OBO (is-a USER) (role concrete) (pattern-match reactive)  
   (slot a (type INTEGER) (range 11 14) (create-accessor ?NONE)))

(defclass LOK (is-a USER) (role concrete) (pattern-match reactive)  
   (slot a (type INTEGER) (range 6 12) (create-accessor ?NONE)))
   
(defclass VOD (is-a USER) (role concrete) (pattern-match reactive) 
   (slot a (type INTEGER) (range 8 10) (create-accessor ?NONE)))

(defrule error-5
  (object (is-a YAK) (a ?x))
  (object (is-a OBO) (a ?x))
  =>)

(defrule error-6
  (object (is-a YAK) (a ?x))
  (object (is-a LOK) (a ?x))
  (object (is-a OBO) (a ?x))
  =>)
  
(defrule error-7
  (object (is-a YAK) (a ?x))
  (object (is-a OBO) (a ?y))
  (object (is-a VOD) (a ?x | ?y))
  =>)

(defrule okay!-2 
  (object (is-a YAK) (a ?x))
  (object (is-a OBO) (a ?y))
  (object (is-a LOK) (a ?x | ?y))
  =>)
  
(defclass KJI (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL) (allowed-symbols on) (create-accessor ?NONE))
  (slot y (type SYMBOL) (allowed-symbols on off) (create-accessor ?NONE)))
               
(defrule error-8
  (object (is-a KJI) (x ?x) (y ?x&~on))
  =>)

(defclass FOO7 (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL FLOAT INTEGER)
           (allowed-symbols d e f) (create-accessor ?NONE)))

(defclass BAR7 (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL FLOAT INTEGER)
           (allowed-symbols a b c) (create-accessor ?NONE)))

(defclass YAK7 (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL)
           (allowed-symbols c d e) (create-accessor ?NONE)))

(defrule okay!-3
  (object (is-a FOO7) (x ?x))
  (object (is-a BAR7) (x ?x))
  =>)
  
(defrule error-9
  (object (is-a FOO7) (x ?x))
  (object (is-a BAR7) (x ?x))
  (object (is-a YAK7) (x ?x))
  =>)

(defclass XXX (is-a USER) (role concrete) (pattern-match reactive)
   (slot a (type INTEGER) (range 6 12) (create-accessor ?NONE)))
   
(defclass YYY (is-a USER) (role concrete) (pattern-match reactive)
   (slot a (type INTEGER) (range 10 16) (create-accessor ?NONE)))
    
(defrule okay!-4
  (object (is-a XXX) (a ?x))
  (object (is-a YYY) (a ?x)) 
  =>)
  
(defrule okay!-5 
  (object (is-a XXX) (a ?x)) 
  (object (is-a YYY) (a ?x)) 
  =>)

;;; Slot x has a minimum of 7 fields and slot q has
;;; a maximum of 5 fields. Binding a variable in both
;;; patterns causes a cardinality constraint violation.

(defclass XOO (is-a USER) (role concrete) (pattern-match reactive)
  (multislot x (cardinality 7 ?VARIABLE) (create-accessor ?NONE)))

(defclass YXK (is-a USER) (role concrete) (pattern-match reactive)
  (multislot q (cardinality ?VARIABLE 5) (create-accessor ?NONE)))

(defrule error-10
  (object (is-a XOO) (x $?x))
  (object (is-a YXK) (q $?x))
  =>)

;;; Slot z has a minimum of 7 fields and slots x and y
;;; a maximum together of 5 fields. A cardinality
;;; violation is caused because $?x and $?y can add
;;; up at most to be 5 fields when placed in the z slot. 

(defclass FOW (is-a USER) (role concrete) (pattern-match reactive)
  (multislot x (cardinality ?VARIABLE 3) (create-accessor ?NONE)))
(defclass BWW (is-a USER) (role concrete) (pattern-match reactive)
  (multislot y (cardinality ?VARIABLE 2) (create-accessor ?NONE)))
(defclass WAK (is-a USER) (role concrete) (pattern-match reactive)
  (multislot z (cardinality 7 ?VARIABLE) (create-accessor ?NONE)))

(defrule error-11
   (object (is-a FOW) (x $?x))
   (object (is-a BWW) (y $?y))
   (object (is-a WAK) (z $?x $?y))
   =>)

;;; More cardinality checks... 

(defclass FLL (is-a USER) (role concrete) (pattern-match reactive)
  (multislot x (cardinality 2 ?VARIABLE) (create-accessor ?NONE)))
(defclass BLR (is-a USER) (role concrete) (pattern-match reactive)
  (multislot y (cardinality 1 ?VARIABLE) (create-accessor ?NONE)))
(defclass QLZ (is-a USER) (role concrete) (pattern-match reactive)
  (multislot z (cardinality 1 ?VARIABLE) (create-accessor ?NONE)))
(defclass YLK (is-a USER) (role concrete) (pattern-match reactive)
  (multislot q (cardinality ?VARIABLE 5) (create-accessor ?NONE)))

(defrule error-12
  (object (is-a FLL) (x $?x))
  (object (is-a BLR) (y $?y))
  (object (is-a QLZ) (z $?z))
  (object (is-a YLK) (q $?x ? $?y ? b $?z))
  =>)

(defrule error-13
  (object (is-a FLL) (x ?))
  =>)

(defrule okay!-6
  (object (is-a FLL) (x ? ?))
  =>)

;;;
;;; ORing of fields
;;;

(defclass MNJ (is-a USER) (role concrete) (pattern-match reactive)
   (slot x (type SYMBOL) (create-accessor ?NONE))
   (slot y (type FLOAT) (create-accessor ?NONE)))

(defrule okay!-7
  (object (is-a MNJ) (x ?x) (y ?y))
  (object (is-a MNJ) (x ?x | ?y) (y ?x | ?y))
  =>)
  
;;;
;;; Check Return Value Types
;;;

(defclass KBP (is-a USER) (role concrete) (pattern-match reactive)
   (slot x (type SYMBOL) (create-accessor ?NONE))
   (slot y (type FLOAT) (create-accessor ?NONE)))
   
(defrule error-14
   (object (is-a KBP) (y ?y) (x =(+ ?y 3)))
   =>)

;;;
;;; Check Types of Variables passed to LHS functions
;;;

(defclass ZC8 (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL) (create-accessor ?NONE))
  (slot y (create-accessor ?NONE)))

(defrule error-15 "Type Violation to LHS function call"
   (object (is-a ZC8) (x ?x) 
        (y ?y&:(> ?y ?x)))
   =>)

(defrule error-16 "Type Violation to LHS function call"
   (object (is-a ZC8) (x ?x) 
        (y ?y))
   (test (> ?y ?x))
   =>)
   
;;;
;;; Check for RHS function conflicts
;;;

(defclass RTY (is-a USER) (role concrete) (pattern-match reactive)
  (slot x (type SYMBOL) (create-accessor ?NONE)))

(defrule error-17
  (object (is-a RTY) (x ?q))
  =>
  (+ 3 5 ?q))
  
(defrule okay!-8 
  (object (is-a RTY) (x ?q))
  =>
  (+ 3 5 ?q)
  (bind ?q 3))
  
(defrule error-18
  (object (is-a RTY))
  =>
  (bind ?q (gensym))
  (+ 3 5 ?q))
  
(defrule okay!-9
  (object (is-a RTY))
  =>
  (bind ?q (gensym))
  (+ 3 5 ?q)
  (bind ?q (* 9 3)))
  
;;;
;;; Yet another test
;;;

(defclass JHJ (is-a USER) (role concrete) (pattern-match reactive)
   (slot x (create-accessor ?NONE))
   (slot y (type SYMBOL) (create-accessor ?NONE)))

(defrule error-19
  (object (is-a JHJ) (x ?x&2) (y ?x))
  =>)

;;;
;;; Constraint Propogation from Expressions
;;;

(defclass YUCK (is-a USER) (role concrete) (pattern-match reactive)
  (slot foo (type SYMBOL) (create-accessor ?NONE)))

(defrule error-20
  (blah ?x&:(> ?x 5))
  (object (is-a YUCK) (foo ?x))
  =>)
  
(defrule error-21
  (object (is-a YUCK) (foo ?x))
  (test (> ?x 5))
  =>)

(defrule error-22
  (blah ?x)
  (test (> ?x 5))
  (object (is-a YUCK) (foo ?x))
  =>)

(defclass BASE (is-a USER) (role abstract)
  (multislot foo
     (type INTEGER)
     (cardinality 3 4)
     (allowed-integers 1 2 3) (create-accessor ?NONE)))

(defclass DERIVE1 (is-a BASE)
  (single-slot foo (source composite)
            (propagation no-inherit) (create-accessor ?NONE)))

(defclass DERIVE2 (is-a DERIVE1)
  (slot foo (type SYMBOL) (source composite) (create-accessor ?NONE)))

(defclass DERIVE2 (is-a DERIVE1)
  (slot foo (type SYMBOL INTEGER) (source composite) (create-accessor ?NONE)))

(defclass DERIVE3 (is-a DERIVE2)
  (slot foo (type SYMBOL INTEGER) (source composite) (cardinality 5 6) (create-accessor ?NONE)))

