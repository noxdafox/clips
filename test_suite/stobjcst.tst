(unwatch all)
(clear)
(set-static-constraint-checking TRUE)
(dribble-on "Actual//stobjcst.out")
(batch "stobjcst.bat")
(dribble-off)
(clear)
(open "Results//stobjcst.rsl" stobjcst "w")
(load "compline.clp")
(printout stobjcst "stobjcst.clp differences are as follows:" crlf)
(compare-files "Expected//stobjcst.out" "Actual//stobjcst.out" stobjcst)
(close stobjcst)
