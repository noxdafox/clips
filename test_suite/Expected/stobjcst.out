TRUE
CLIPS> (batch "stobjcst.bat")
TRUE
CLIPS> (load "stobjcst.clp")
##*
[RULECSTR1] Variable ?x in CE #1 slot y
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-1
   (object (is-a FOO)
           (x ?x)
           (y ?x))
   =>)
*
[RULECSTR1] Variable ?x in CE #2 slot y
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-2
   (object (is-a FOO)
           (x ?x))
   (object (is-a FOO)
           (y ?x))
   =>)
*
[RULECSTR1] Variable ?x in CE #1 slot c
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-3
   (object (is-a BAR)
           (a ?x)
           (b ?x)
           (c ?x))
   =>)
*
[RULECSTR1] Variable ?x in CE #3 slot c
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-4
   (object (is-a BAR)
           (a ?x))
   (object (is-a BAR)
           (b ?x))
   (object (is-a BAR)
           (c ?x))
   =>)
*####*
[RULECSTR1] Variable ?x in CE #2 slot a
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-5
   (object (is-a YAK)
           (a ?x))
   (object (is-a OBO)
           (a ?x))
   =>)
*
[RULECSTR1] Variable ?x in CE #3 slot a
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-6
   (object (is-a YAK)
           (a ?x))
   (object (is-a LOK)
           (a ?x))
   (object (is-a OBO)
           (a ?x))
   =>)
*
[RULECSTR1] Pattern #3 slot a
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-7
   (object (is-a YAK)
           (a ?x))
   (object (is-a OBO)
           (a ?y))
   (object (is-a VOD)
           (a ?x|?y))
   =>)
*#*
[RULECSTR1] Variable ?x in CE #1 slot y
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-8
   (object (is-a KJI)
           (x ?x)
           (y ?x&~on))
   =>)
###**
[RULECSTR1] Variable ?x in CE #3 slot x
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-9
   (object (is-a FOO7)
           (x ?x))
   (object (is-a BAR7)
           (x ?x))
   (object (is-a YAK7)
           (x ?x))
   =>)
##**##*
[RULECSTR1] Variable ?x in CE #2 slot q
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-10
   (object (is-a XOO)
           (x $?x))
   (object (is-a YXK)
           (q $?x))
   =>)
###*
[CSTRNCHK1] The group of restrictions found in CE #3
does not satisfy the cardinality restrictions for slot z.

ERROR:
(defrule MAIN::error-11
   (object (is-a FOW)
           (x $?x))
   (object (is-a BWW)
           (y $?y))
   (object (is-a WAK)
           (z $?x $?y))
   =>)
####*
[CSTRNCHK1] The group of restrictions found in CE #4
does not satisfy the cardinality restrictions for slot q.

ERROR:
(defrule MAIN::error-12
   (object (is-a FLL)
           (x $?x))
   (object (is-a BLR)
           (y $?y))
   (object (is-a QLZ)
           (z $?z))
   (object (is-a YLK)
           (q $?x ? $?y ? b $?z))
   =>)
*
[CSTRNCHK1] The group of restrictions found in CE #1
does not satisfy the cardinality restrictions for slot x.

ERROR:
(defrule MAIN::error-13
   (object (is-a FLL)
           (x ?))
   =>)
*#*#*
[RULECSTR1] Pattern #1 slot x
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-14
   (object (is-a KBP)
           (y ?y)
           (x =(+ ?y 3)))
   =>)
#*
[RULECSTR2] Previous variable bindings of ?x caused the type restrictions
for argument #2 of the expression (> ?y ?x)
found in CE #1 slot y to be violated.

ERROR:
(defrule MAIN::error-15 "Type Violation to LHS function call"
   (object (is-a ZC8)
           (x ?x)
           (y ?y&:(> ?y ?x)))
   =>)
*
[RULECSTR2] Previous variable bindings of ?x caused the type restrictions
for argument #2 of the expression (> ?y ?x)
found in CE #2 to be violated.

ERROR:
(defrule MAIN::error-16 "Type Violation to LHS function call"
   (object (is-a ZC8)
           (x ?x)
           (y ?y))
   (test (> ?y ?x))
   =>)
#*
[RULECSTR3] Previous variable bindings of ?q caused the type restrictions
for argument #3 of the expression (+ 3 5 ?q)
found in the rule's RHS to be violated.

ERROR:
(defrule MAIN::error-17
   (object (is-a RTY)
           (x ?q))
   =>
   (+ 3 5 ?q))
**
[RULECSTR3] Previous variable bindings of ?q caused the type restrictions
for argument #3 of the expression (+ 3 5 ?q)
found in the rule's RHS to be violated.

ERROR:
(defrule MAIN::error-18
   (object (is-a RTY))
   =>
   (bind ?q (gensym))
   (+ 3 5 ?q))
*#*
[RULECSTR1] Variable ?x in CE #1 slot y
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-19
   (object (is-a JHJ)
           (x ?x&2)
           (y ?x))
   =>)
#*
[RULECSTR1] Variable ?x in CE #2 slot foo
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-20
   (blah ?x&:(> ?x 5))
   (object (is-a YUCK)
           (foo ?x))
   =>)
*
[RULECSTR2] Previous variable bindings of ?x caused the type restrictions
for argument #1 of the expression (> ?x 5)
found in CE #2 to be violated.

ERROR:
(defrule MAIN::error-21
   (object (is-a YUCK)
           (foo ?x))
   (test (> ?x 5))
   =>)
*
[RULECSTR1] Variable ?x in CE #3 slot foo
has constraint conflicts which make the pattern unmatchable.

ERROR:
(defrule MAIN::error-22
   (blah ?x)
   (test (> ?x 5))
   (object (is-a YUCK)
           (foo ?x))
   =>)
###
[CSTRNPSR1] The type attribute conflicts with the allowed-integers/numbers attribute.

ERROR:
(defclass MAIN::DERIVE2
   (is-a DERIVE1)
   (slot foo
      (type SYMBOL)
      (source composite)
      (create-accessor ?NONE))
##
FALSE
CLIPS> (list-defrules)
okay!-1
okay!-2
okay!-3
okay!-4
okay!-5
okay!-6
okay!-7
okay!-8
okay!-9
For a total of 9 defrules.
CLIPS> (describe-class BASE)
================================================================================
********************************************************************************
Abstract: direct instances of this class cannot be created.

Direct Superclasses: USER
Inheritance Precedence: BASE USER OBJECT
Direct Subclasses: DERIVE1
--------------------------------------------------------------------------------
SLOTS : FLD DEF PRP ACC STO MCH SRC VIS CRT OVRD-MSG SOURCE(S)
foo   : MLT STC INH RW  LCL RCT EXC PRV NIL put-foo  BASE

Constraint information for slots:

SLOTS : SYM STR INN INA EXA FTA INT FLT
foo   :                          #      RNG:[-oo..+oo] CRD:[3..4]
--------------------------------------------------------------------------------
Recognized message-handlers:
init primary in class USER
delete primary in class USER
create primary in class USER
print primary in class USER
direct-modify primary in class USER
message-modify primary in class USER
direct-duplicate primary in class USER
message-duplicate primary in class USER
********************************************************************************
================================================================================
CLIPS> (describe-class DERIVE1)
================================================================================
********************************************************************************
Abstract: direct instances of this class cannot be created.

Direct Superclasses: BASE
Inheritance Precedence: DERIVE1 BASE USER OBJECT
Direct Subclasses: DERIVE2
--------------------------------------------------------------------------------
SLOTS : FLD DEF PRP ACC STO MCH SRC VIS CRT OVRD-MSG SOURCE(S)
foo   : SGL STC NIL RW  LCL RCT CMP PRV NIL put-foo  BASE DERIVE1

Constraint information for slots:

SLOTS : SYM STR INN INA EXA FTA INT FLT
foo   :                          #      RNG:[-oo..+oo] 
--------------------------------------------------------------------------------
Recognized message-handlers:
init primary in class USER
delete primary in class USER
create primary in class USER
print primary in class USER
direct-modify primary in class USER
message-modify primary in class USER
direct-duplicate primary in class USER
message-duplicate primary in class USER
********************************************************************************
================================================================================
CLIPS> (describe-class DERIVE2)
================================================================================
********************************************************************************
Abstract: direct instances of this class cannot be created.

Direct Superclasses: DERIVE1
Inheritance Precedence: DERIVE2 DERIVE1 BASE USER OBJECT
Direct Subclasses: DERIVE3
--------------------------------------------------------------------------------
SLOTS : FLD DEF PRP ACC STO MCH SRC VIS CRT OVRD-MSG SOURCE(S)
foo   : MLT STC INH RW  LCL RCT CMP PRV NIL put-foo  BASE DERIVE2

Constraint information for slots:

SLOTS : SYM STR INN INA EXA FTA INT FLT
foo   :  +                       #      RNG:[-oo..+oo] CRD:[3..4]
--------------------------------------------------------------------------------
Recognized message-handlers:
init primary in class USER
delete primary in class USER
create primary in class USER
print primary in class USER
direct-modify primary in class USER
message-modify primary in class USER
direct-duplicate primary in class USER
message-duplicate primary in class USER
********************************************************************************
================================================================================
CLIPS> (describe-class DERIVE3)
================================================================================
********************************************************************************
Abstract: direct instances of this class cannot be created.

Direct Superclasses: DERIVE2
Inheritance Precedence: DERIVE3 DERIVE2 DERIVE1 BASE USER OBJECT
Direct Subclasses:
--------------------------------------------------------------------------------
SLOTS : FLD DEF PRP ACC STO MCH SRC VIS CRT OVRD-MSG SOURCE(S)
foo   : MLT STC INH RW  LCL RCT CMP PRV NIL put-foo  BASE DERIVE2 DERIVE3

Constraint information for slots:

SLOTS : SYM STR INN INA EXA FTA INT FLT
foo   :  +                       #      RNG:[-oo..+oo] CRD:[5..6]
--------------------------------------------------------------------------------
Recognized message-handlers:
init primary in class USER
delete primary in class USER
create primary in class USER
print primary in class USER
direct-modify primary in class USER
message-modify primary in class USER
direct-duplicate primary in class USER
message-duplicate primary in class USER
********************************************************************************
================================================================================
CLIPS> (clear)
CLIPS> (defclass A (is-a USER)
  (slot foo (cardinality 3 4)))

[CLSLTPSR3] Cardinality facet can only be used with multifield slots

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (cardinality 3 4))
CLIPS> (defclass A (is-a USER)
  (slot foo (type SYMBOL) (range 3 4)))

[CSTRNPSR1] The type attribute conflicts with the range attribute.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (type SYMBOL)
      (range 3 4))
CLIPS> (defclass A (is-a USER)
  (slot foo (min-number-of-elements 3)))

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for defclass slot.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (min-number-of-elements
CLIPS> (defclass A (is-a USER)
  (slot foo (allowed-words abc)))

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for defclass slot.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (allowed-words
CLIPS> (defclass A (is-a USER)
  (slot foo (allowed-values)))

[PRNTUTIL2] Syntax Error:  Check appropriate syntax for allowed-values attribute.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (allowed-values )
CLIPS> (defclass A (is-a USER)
  (slot foo (allowed-integers abc)))

[CSTRNPSR4] Value does not match the expected type for the allowed-integers attribute

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (allowed-integers abc 
CLIPS> (defclass A (is-a USER)
   (slot foo (type INTEGER) (default abc)))
[CSTRNCHK1] abc for slot foo found in slot default value
does not match the allowed types.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (type INTEGER)
      (default abc))
CLIPS> (defclass A (is-a USER)
   (slot foo (type INTEGER) (default-dynamic abc)))
[CSTRNCHK1] Expression for slot foo found in dynamic default value
does not match the allowed types.

ERROR:
(defclass MAIN::A
   (is-a USER)
   (slot foo
      (type INTEGER)
      (default-dynamic abc))
CLIPS> (set-static-constraint-checking FALSE)
TRUE
CLIPS> (defclass A (is-a USER)
   (slot foo (type INTEGER) (default abc)))
CLIPS> (defclass B (is-a USER)
   (slot foo (type INTEGER) (default-dynamic abc)))
CLIPS> (set-static-constraint-checking TRUE)
FALSE
CLIPS> (clear)
CLIPS> (dribble-off)
